File: engine/src/parser.rs
use crate::ast::{Ast, OpCall};
use crate::lexer::{Token, TokenType};
use std::collections::HashMap;

/// StarQL Grammar:
///
/// query :=  statement (";" statement) * ";"?
/// statement := (identifier "=")? op_expression
/// op_expression := expression | ( "|" opcall) *
/// opcall := identifier ("." identifier) * "(" parameters ")"
/// expression := logic_or
/// logic_or := logic_and ( "or" logic_and )*
/// logic_and := equality ( "and" equality )*
/// equality := comparison ( ( "!=" | "==" ) comparison )*
/// comparison := term ( ( ">" | ">=" | "<" | "<=" ) term )*
/// term := factor (("+" | "-") factor)*
/// factor := unary (("*" | "/") unary)*
/// unary := primary | "-" unary
/// symbol := "$" identifier
/// primary := NUMBER | STRING | "true" | "false" | "(" expression ")" | list | record | symbol
/// list := "[" (expression ",")* expression? "]"
/// record := "{" parameters "}"
/// parameters := (identifier "=" expression ",")* (identifier "=" expression)?
///
/// To add:
/// * Booleans
/// * Symbols
/// Plan:
///     add bool, string, number in expr enum
/// TODOs:
///     replace anyhow::Result with our own ParseError error class?
///

// 'a denotes the lifetime of slice (aka lifetime of parser)
// 'b denotes the lifetime of tokens
pub struct Parser<'a> {
    tokens: Vec<Token<'a>>,
    previous: Option<Token<'a>>,
}

impl<'a> Parser<'a> {
    pub fn new(mut tokens: Vec<Token<'a>>) -> Self {
        tokens.reverse();
        Parser {
            tokens: tokens,
            previous: None,
        }
    }

    fn peek(&self) -> Option<&Token<'a>> {
        self.tokens.last()
    }

    fn previous(&mut self) -> Option<Token<'a>> {
        self.previous.take()
    }

    fn done(&self) -> bool {
        if let Some(t) = self.peek() {
            if t.token_type == TokenType::Eof {
                return true;
            }
        }
        return false;
    }

    fn advance(&mut self) {
        if !self.done() {
            self.previous = self.tokens.pop();
        }
    }

    fn matches(&mut self, token_types: &[TokenType]) -> bool {
        for t in token_types {
            if self.check(*t) {
                self.advance();
                return true;
            }
        }
        return false;
    }

    fn check(&self, token_type: TokenType) -> bool {
        if self.done() {
            return false;
        }
        if let Some(token) = self.peek() {
            if token.token_type == token_type {
                return true;
            };
        }
        return false;
    }

    fn consume(&mut self, token_type: TokenType) -> anyhow::Result<Token<'a>> {
        if self.check(token_type) {
            self.advance();
            return self.previous().ok_or_else(|| anyhow::anyhow!("missing"));
        } else {
            return Err(anyhow::anyhow!(
                "Unexpected token: {:?}. Expected token of type: {:?}",
                self.peek(),
                token_type,
            ));
        }
    }

    pub fn parse(&mut self) -> anyhow::Result<Ast<'a>> {
        self.query()
    }

    fn query(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut statements = vec![self.statement()?];
        let semicolon = &[TokenType::Semicolon];
        while self.matches(semicolon) {
            if self.done() {
                break;
            }
            statements.push(self.statement()?);
        }
        // consume optional trailing semi colon
        self.matches(semicolon);
        Ok(Ast::Query(statements))
    }

    fn statement(&mut self) -> anyhow::Result<Ast<'a>> {
        let variable = if let Ok(s) = self.identifier() {
            self.consume(TokenType::Equal)?;
            Some(s)
        } else {
            None
        };
        let opexp = self.op_expression()?;
        Ok(Ast::Statement(variable, Box::new(opexp)))
    }

    fn op_expression(&mut self) -> anyhow::Result<Ast<'a>> {
        let e = self.expression()?;
        let mut opcalls: Vec<OpCall> = vec![];
        let pipe = &[TokenType::Pipe];
        while self.matches(pipe) {
            opcalls.push(self.opcall()?);
        }
        Ok(Ast::OpExp(Box::new(e), opcalls))
    }

    fn opcall(&mut self) -> anyhow::Result<OpCall<'a>> {
        let mut path: Vec<Token> = vec![];
        let dot = &[TokenType::Dot];
        loop {
            path.push(self.identifier()?);
            if !self.matches(dot) {
                break;
            }
        }
        let mut args = HashMap::new();
        self.consume(TokenType::LeftParen)?;
        let comma = &[TokenType::Comma];
        loop {
            let k = self.identifier()?;
            self.consume(TokenType::Equal)?;
            let e = self.expression()?;
            args.insert(k, e);
            if !self.matches(comma) {
                break;
            }
            if self.check(TokenType::RightParen) {
                break;
            }
        }
        self.consume(TokenType::RightParen)?;
        Ok(OpCall { path, args })
    }

    fn identifier(&mut self) -> anyhow::Result<Token<'a>> {
        self.consume(TokenType::Identifier)
    }

    fn expression(&mut self) -> anyhow::Result<Ast<'a>> {
        self.logic_or()
    }

    fn term(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut f = self.factor()?;
        let expected = &[TokenType::Plus, TokenType::Minus];
        while self.matches(expected) {
            let op = self.previous().unwrap();
            let right = self.factor()?;
            f = Ast::Binary {
                left: Box::new(f),
                op: op,
                right: Box::new(right),
            }
        }
        Ok(f)
    }

    fn factor(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut u = self.unary()?;
        let expected = &[TokenType::Star, TokenType::Slash];
        while self.matches(expected) {
            let op = self.previous().unwrap();
            let right = self.unary()?;
            u = Ast::Binary {
                left: Box::new(u),
                op: op,
                right: Box::new(right),
            }
        }
        Ok(u)
    }

    fn unary(&mut self) -> anyhow::Result<Ast<'a>> {
        let ops = &[TokenType::Minus];
        if self.matches(ops) {
            let op = self.previous().unwrap();
            let right = self.unary()?;
            Ok(Ast::Unary(op, Box::new(right)))
        } else {
            self.primary()
        }
    }

    fn list(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut l = vec![];
        let comma = &[TokenType::Comma];
        while !self.check(TokenType::ListEnd) {
            let e = self.expression()?;
            l.push(e);
            if !self.matches(comma) {
                break;
            }
        }
        self.consume(TokenType::ListEnd)?;
        Ok(Ast::List(l))
    }

    fn record(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut r = HashMap::new();
        let comma = &[TokenType::Comma];
        while !self.check(TokenType::RecordEnd) {
            let id = self.identifier()?;
            self.consume(TokenType::Equal)?;
            let e = self.expression()?;
            r.insert(id, e);
            if !self.matches(comma) {
                break;
            }
        }
        self.consume(TokenType::RecordEnd)?;
        Ok(Ast::Record(r))
    }

    fn primary(&mut self) -> anyhow::Result<Ast<'a>> {
        if self.matches(&vec![
            TokenType::Number,
            TokenType::String,
            TokenType::Bool,
            TokenType::Variable,
            TokenType::Identifier,
        ]) {
            Ok(Ast::Atom(self.previous().unwrap()))
        } else if self.matches(&[TokenType::LeftParen]) {
            let e = self.expression();
            self.consume(TokenType::RightParen)?;
            e
        } else if self.matches(&[TokenType::ListBegin]) {
            self.list()
        } else if self.matches(&[TokenType::RecordBegin]) {
            self.record()
        } else {
            anyhow::bail!("Unexpected token: {:?}", self.peek())
        }
    }

    fn comparison(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut f = self.term()?;
        let expected = vec![
            TokenType::Greater,
            TokenType::GreaterEqual,
            TokenType::Lesser,
            TokenType::LesserEqual,
        ];
        while self.matches(&expected) {
            let op = self.previous().unwrap();
            let right = self.term()?;
            f = Ast::Binary {
                op: op,
                left: Box::new(f),
                right: Box::new(right),
            }
        }
        Ok(f)
    }
    fn equality(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut f = self.comparison()?;
        let expected = &[TokenType::EqualEqual, TokenType::BangEqual];
        while self.matches(expected) {
            let op = self.previous().unwrap();
            let right = self.comparison()?;
            f = Ast::Binary {
                op: op,
                left: Box::new(f),
                right: Box::new(right),
            }
        }
        Ok(f)
    }
    fn logic_and(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut f = self.equality()?;
        while self.matches(&[TokenType::And]) {
            let op = self.previous().unwrap();
            let right = self.equality()?;
            f = Ast::Binary {
                op: op,
                left: Box::new(f),
                right: Box::new(right),
            }
        }
        Ok(f)
    }
    fn logic_or(&mut self) -> anyhow::Result<Ast<'a>> {
        let mut f = self.logic_and()?;
        while self.matches(&[TokenType::Or]) {
            let op = self.previous().unwrap();
            let right = self.logic_and()?;
            f = Ast::Binary {
                op: op,
                left: Box::new(f),
                right: Box::new(right),
            }
        }
        Ok(f)
    }
}

File: kengine/src/main/kotlin/starql/parser/Parser.kt
package starql.parser

import starql.ParseException
import starql.ast.*
import starql.ast.List
import starql.lexer.Lexer
import starql.lexer.Token
import starql.lexer.TokenType

/*
query          → statement +
statement      → (identifier "=")? expression ";"
expression     → logic_or ("|" operator)*
operator       → identifier "." identifier "("parameter ("," parameter)* ","? ")"

logic_or       → logic_and ( "or" logic_and )*
logic_and      → equality ( "and" equality )*
equality       → comparison ( ( "!=" | "==" ) comparison )*
comparison     → term ( ( ">" | ">=" | "<" | "<=" ) term )*

term           → factor ( ( "-" | "+" ) factor )*
factor         → unary ( ( "/" | "*" ) unary )*
unary          → ( "!" | "-" ) primary | primary
primary        → "true" | "false" | Number | String | list | dict | "(" expression ")" | variable | table
list           → "[" expression ("," expression)* ","? "]"
dict           → "{" identifier "=" expression ("," identifier "=" expression)* ","? "]"
variable       → ("$" identifier | "@") lookup*
lookup         → ("." identifier) |  "[" expression "]"

 */

class Parser(private val query: String) {
    private val lexer = Lexer(query)
    private var current: Token? = null
    private var next: Token = lexer.next()
    private val varsSoFar = HashSet<String>()
    private val errors = arrayListOf<ParseException>()
    private var needsRecovery = false

    private fun advance(): Token {
        current = next
        next = lexer.next()
        return next
    }

    private fun error(error: ParseException) {
        errors.add(error)
        needsRecovery = true
    }

    private fun recover() {
        // consumes some tokens until the next statement to fix the internal state of parser
        while (current!!.type !in listOf(TokenType.Eof, TokenType.Semicolon)) {
            advance()
        }
    }

    private fun matches(vararg types: TokenType): Boolean {
        return if (next.type in types) {
            advance()
            true
        } else {
            false
        }
    }

    private fun expect(vararg types: TokenType): Token {
        return if (next.type in types) {
            advance()
            current!!
        } else {
            throw ParseException("unexpected token '$next'", current, types.map { it.toString() })
        }
    }

    private fun term(): Ast {
        var l = factor()
        while (matches(TokenType.Plus, TokenType.Minus)) {
            val op = current!!
            val r = factor()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun unary(): Ast {
        return if (matches(TokenType.Bang, TokenType.Minus)) {
            val op = current!!
            val v = primary()
            Unary(op, v)
        } else {
            primary()
        }
    }

    private fun factor(): Ast {
        var l = unary()
        while (matches(TokenType.Star, TokenType.Slash)) {
            val op = current!!
            val r = unary()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun primary(): Ast {
        advance()
        return when (current!!.type) {
            TokenType.Number, TokenType.Bool, TokenType.String -> Atom(current!!)
            TokenType.ListBegin -> list(true)
            TokenType.RecordBegin -> dict(true)
            TokenType.LeftParen -> grouping(true)
            TokenType.Dollar -> variable(true)
            TokenType.At -> variable(true)
            TokenType.Table -> table(true)
            else -> throw ParseException(
                "unexpected token '$current'", current,
                listOf("digit", "[", "{", "(", "$", "@", "table")
            )
        }
    }


    private fun list(prefixDone: Boolean): Ast {
        if (!prefixDone) {
            expect(TokenType.ListBegin)
        }
        val elements = arrayListOf<Ast>()
        while (next.type != TokenType.ListEnd) {
            elements.add(expression())
            when (next.type) {
                TokenType.ListEnd -> break
                TokenType.Comma -> advance()
                else -> throw ParseException("unexpected token '$next'", current, listOf(",", "]"))
            }
        }
        expect(TokenType.ListEnd)
        return List(elements)
    }

    private fun dict(prefixDone: Boolean): Ast {
        if (!prefixDone) {
            expect(TokenType.RecordBegin)
        }
        val elements = HashMap<Token, Ast>()
        while (next.type != TokenType.RecordEnd) {
            val (identifier, exp) = parameter()
            elements[identifier] = exp
            when (next.type) {
                TokenType.RecordEnd -> break
                TokenType.Comma -> advance()
                else -> throw ParseException("unexpected token '$next'", current, listOf(",", "}"))
            }
        }
        expect(TokenType.RecordEnd)
        return Dict(elements)
    }

    private fun parameter(): Pair<Token, Ast> {
        val p = expect(TokenType.Identifier)
        expect(TokenType.Equal)
        val e = expression()
        return Pair<Token, Ast>(p, e)
    }

    private fun or(): Ast {
        var l = and()
        while (matches(TokenType.Or)) {
            val op = current!!
            val r = and()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun and(): Ast {
        var l = equality()
        while (matches(TokenType.And)) {
            val op = current!!
            val r = equality()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun comparison(): Ast {
        var l = term()
        while (matches(TokenType.Greater, TokenType.GreaterEqual, TokenType.Lesser, TokenType.LesserEqual)) {
            val op = current!!
            val r = term()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun grouping(prefixDone: Boolean): Ast {
        if (!prefixDone) {
            expect(TokenType.LeftParen)
        }
        val ret = expression()
        expect(TokenType.RightParen)
        return Grouping(ret)
    }

    private fun equality(): Ast {
        var l = comparison()
        while (matches(TokenType.BangEqual, TokenType.EqualEqual)) {
            val op = current!!
            val r = comparison()
            l = Binary(l, op, r)
        }
        return l
    }

    private fun expression(): Ast {
        var e = or()
        while (matches(TokenType.Pipe)) {
            val module = expect(TokenType.Identifier)
            expect(TokenType.Dot)
            val name = expect(TokenType.Identifier)
            expect(TokenType.LeftParen)
            val args = hashMapOf<Token, Ast>()
            while (next.type != TokenType.RightParen) {
                val (key, exp) = parameter()
                args[key] = exp
                when (next.type) {
                    TokenType.RightParen -> break
                    TokenType.Comma -> advance()
                    else -> throw ParseException("unexpected token '$next'", current, listOf(",", ")"))
                }
            }
            expect(TokenType.RightParen)
            e = Opcall(e, module, name, args)
        }
        return e
    }

    private fun variable(prefixDone: Boolean): Ast {
        if (!prefixDone) {
            expect(TokenType.At, TokenType.Dollar)
        }
        val v = when (current!!.type) {
            TokenType.Dollar -> {
                val name = expect(TokenType.Identifier)
                if (name.literal() !in varsSoFar) {
                    throw ParseException(
                        "referring to undefined variable: '${name.literal()}'",
                        current,
                        listOf<String>()
                    )
                }
                name
            }
            TokenType.At -> current!!
            else -> throw ParseException("unexpected token '$current'", current, listOf("@", "$"))
        }

        val lookups = ArrayList<Ast>()
        while (true) {
            when (next.type) {
                TokenType.Dot -> {
                    advance()
                    val id = expect(TokenType.Identifier)
                    lookups.add(Atom(id))
                }
                TokenType.ListBegin -> {
                    advance()
                    val e = expression()
                    expect(TokenType.ListEnd)
                    lookups.add(e)
                }
                else -> break
            }
        }
        return Var(v, lookups)
    }

    private fun statement(): Ast {
        var name: Token? = null
        if (next.type == TokenType.Identifier) {
            advance()
            name = current
            expect(TokenType.Equal)
            varsSoFar.add(name!!.literal())
        }
        val body = expression()
        expect(TokenType.Semicolon)

        return Statement(name, body)
    }

    private fun query(): Ast {
        val statements = ArrayList<Ast>(listOf(statement()))
        while (!matches(TokenType.Eof)) {
            statements.add(statement())
        }
        return Query(statements)
    }

    private fun table(prefixDone: Boolean): Ast {
        if (!prefixDone) {
            expect(TokenType.Table)
        }
        expect(TokenType.LeftParen)
        val e = expression()
        expect(TokenType.RightParen)
        return Table(e)
    }

    fun parse(): Ast {
        val statements = ArrayList<Ast>()
        while (!matches(TokenType.Eof)) {
            try {
                if (needsRecovery) {
                    recover()
                    needsRecovery = false
                }
                statements.add(statement())
            } catch (error: ParseException) {
                error(error)
            }
        }
        if (errors.isEmpty() && statements.isEmpty()) {
            error(ParseException("zero valid statements in query", current, listOf()))
        }
        if (errors.isNotEmpty()) {
            for (e in errors) {
                println(e)
            }
            throw errors.first()
        }
        return Query(statements)
    }
}